src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.fetch;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.BitSet;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.lucene.search.Queries;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.common.bytes.BytesReference;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.IndexSettings;
+import org.opensearch.index.mapper.DocValueFetcher;
+import org.opensearch.index.mapper.DocumentMapper;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.ObjectMapper;
+import org.opensearch.index.mapper.ValueFetcher;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.mapper.KNNVectorFieldMapper;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.fetch.FetchContext;
+import org.opensearch.search.fetch.FetchSubPhase;
+import org.opensearch.search.fetch.FetchSubPhaseProcessor;
+import org.opensearch.search.internal.ContextIndexSearcher;
+import org.opensearch.search.lookup.SourceLookup;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.knn.common.KNNConstants.BYTES_PER_KILOBYTES;
+
+/**
+ * Fetch sub phase which pull data from doc values.
+ * and fulfill the value into source map
+ */
+@Log4j2
+public class KNNFetchSubPhase implements FetchSubPhase {
+
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) throws IOException {
```

Comment:
@bugmakerrrrrr I thought of another usecase which I want to know how we are handling here. 

So, in the search request I can mention that in `_source` don't retrieve XYZ field. 

If a user has enabled this for vector field will we still fetch the vectors from doc values? Can you please check this use case too. Here is a sample query:

```
POST embedding-index/_search
{
  "_source": {
    "excludes": "my_vector1"
  },
  "query": {
    "knn": {
      "my_vector1": {
        "k": 10,
        "vector": [1,2,3,4]
      }
    }
  }
}
```
In my understanding with current code we will still fetch the doc values for vectors.

Reply:
> If a user has enabled this for vector field will we still fetch the vectors from doc values? Can you please check this use case too. Here is a sample query:

@navneet1v 

- when we enable the source, we can disabled the settings `index.knn.synthetic_source.enabled` which defaults value is false and it would not go into the `KNNFetchSubPhase#process`. and get the source directly

- when we enable the source, and enabled the `synthetic_source` settings, it go into the `KNNFetchSubPhase#process` and as the following codes checked SourceAsMap which would contains the source field, and skip fetch value from docValues

https://github.com/opensearch-project/k-NN/blob/3c7f0e328b00ea418c86020b41c701b9af8d784b/src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java#L111-L119

Reply:
@navneet1v @luyuncheng I just went over this PR, please correct me if I have some misunderstandings.

> In my understanding with current code we will still fetch the doc values for vectors.

Yes, if we exclude the vector field in index mappings and enable the synthetic_source setting, it will be a problem when we use the Lucene engine and disable the doc value of the vector field. To address this issue, we can fetch values from the `Byte/FloatVectorValues`, just like I did in `ScriptDocValues`. I can submit a PR after this PR merged, and then I will submit another PR to disable the doc values totally for the Lecene engine.

Reply:
@luyuncheng my question here is not about if the vector field is not present in source then `KNNFetchSubPhase` will run or not. My question was lets say in the mappings I don't exclude the vector field, but I can exclude the fetching of vector field during my search query like this:

```
POST embedding-index/_search
{
  "_source": {
    "excluded": "my_vector1"
  },
  "query": {
    "knn": {
      "my_vector1": {
        "k": 10,
        "vector": [1,2,3,4]
      }
    }
  }
}
```

My worry is as the FetchSourcePhase will not add the vector field in the response but `KNNFetchSubPhase` will add it in the Search Response. My suggestion is we should read this information from search context and see if the vector field is marked as excluded or not before even fetching the Vector via DocValues.